packages/teuchos/comm/src/Teuchos_Comm.hpp

```diff
@@ -662,6 +718,34 @@ class Comm : virtual public Describable {
   virtual RCP<CommStatus<Ordinal> >
   wait (const Ptr<RCP<CommRequest<Ordinal> > >& request) const = 0;
 
+  /// \brief Start a single communication request.
+  ///
+  /// \param request [in/out] On input: request is not null, and
+  /// <tt>*request</tt> is either null (in which case this function
+  /// does nothing and returns null) or an RCP of a valid CommRequest
+  /// instance representing an outstanding communication request.
+  virtual void
+  start (const Ptr<RCP<CommRequest<Ordinal> > >& request) const = 0;
+
+  /** \brief Start on a set of communication requests.
+   *
+   * <b>Preconditions:</b><ul>
+   * <li> <tt>requests.size() > 0</tt>
+   * </ul>
+   *
+   */
+  virtual void
+  startAll(const ArrayView<RCP<CommRequest<Ordinal> > > &requests) const = 0;
+
+  /// \brief Free a single communication request.
+  ///
+  /// \param request [in/out] On input: request is not null, and
+  /// <tt>*request</tt> is either null (in which case this function
+  /// does nothing and returns null) or an RCP of a valid CommRequest
+  /// instance representing an communication request.
+  virtual void
+  free (const Ptr<RCP<CommRequest<Ordinal> > >& request) const = 0;
```

Comment:
Do you actually need an explicit `free` method?  CommRequest objects know how to free themselves already.

Reply:
The `MpiCommRequestBase` has a destructor that calls `MPI_Cancel`, not `MPI_Request_free`. I think that is is reasonable in the case of persistent requests to ask the user to explicitly free them..

Reply:
@cgcgcg Then we should fix the destructor :-)